### PR TITLE
Model Submission for Marble HD2 sensor configuration 1

### PR DESCRIPTION
**Pull request by nkoenig**
Note: This model requires a custom ros_ign_bridge to be built (or our latest PR for ros_ign_bridge to be incorporated). Instructions for where to find the source code can be found in “submitted_models/marble_hd2_sensor_config_1/BuildingCustomRosIgnBridgeForMarble.txt”

See specifications.md file for more information

Sensors:
* camera_front (based on Intel RealSense D435)
  * 1280 x 720 resolution
  * Depth range of 10m to 0.1m
  * 30 fps
  * 86 degree Horizontal FOV and 57 degree vertical FOV
  * Topics: vehicle_name/front_facing/image_raw & vehicle_name/front_facing/depth & vehicle_name/front_facing_rgbd_camera/depth/points & vehicle_name/front_facing/etc…
* camera_down (based on Intel RealSense D435)
  * Same specifications as camera_front
  * Topics: vehicle_name/front_down/image_raw & vehicle_name/front_down/depth & vehicle_name/front_down_rgbd_camera/depth/points & vehicle_name/front_down/etc…
* camera_pan_tilt (based on Intel RealSense D435)
  * Same specifications as camera_front and camera_down
  * Topics: vehicle_name/pan_tilt/image_raw & vehicle_name/pan_tilt/depth & vehicle_name/pan_tilt_rgbd_camera/depth/points & vehicle_name/pan_tilt/etc…
* planar_laser (Based off RPLidar S1)
  * 10 hz
  * 920 horizontal samples
  * 360 degree horizontal range
  * 0.04m to 25m range
  * 3cm resolution
  * Topics: vehicle_name/planar_laser_scan 
* front_laser (Based off OS1-64 Laser in 20x1024 configuration)
  * 1024 horizontal samples
  * 360 degree horizontal range
  * 64 vertical samples
  * 45 degree vertical range
  * 0.09m to 120m range
  * 1cm resolution
  * Topics: vehicle_name/points
* IMU (Based off Lord Microstrain 3DM-GX5-25 IMU)
  * https://www.microstrain.com/sites/default/files/applications/files/3dm-gx5-25_datasheet_8400-0093_rev_n.pdf
* Pan-Tilt Gimbal
  * Pan range: -180 to 180 degrees
  * Tilt Range: -160 to 160 degrees
  * Topics: vehicle_name/pan_tilt/pan_rate_cmd_double & vehicle_name/pan_tilt/tilt_rate_cmd_double

### CODE REVIEW

**Comment by acschang** [comment]
The initial assessment of the MARBLE HD2 model is complete, however there are some issues to resolve before this model can be merged and utilized in a competitive setting:

- The RPLIDAR seems to be oriented at an off-angle and returns ground strikes when the robot is positioned on a flat surface.  Is this intentional?
- Please change the IMU noise model to match the parameters of the X1-X4 models.
- The mass listed in `specifications.md` does not match the mass in `model.sdf`.  Please correct the mass in the `model.sdf` file.
- The gimbal seems to be non-functional (topic is not present to command, and the included script `test_gimbal.sh` did not work). Can you double-check the implementation?
- The validation data is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.

Please address the issues noted above and then submit a pull request to update your submission by **5 JUNE**. The submissions are expected to be incorporated into the SubT Virtual Testbed pending a successful review.

**Comment by bfotheri** [comment]
Thanks for the comments. I apologize that we didn't see this until now. Perhaps we need to reconfigure our gits account notification settings because we didn't see this comment until we actually looked at the PR's on the osrf repo one by one. We have a number of small changes we've been working on for the past 2 months in our local branches that we would still like to submit a pull request for if possible. 

RPLIDAR: I wasn't able to reproduce the RPLIDAR getting ground strikes (the model sdf shows it having a (0, -0, 0) RPY) but it does self intersect and so we're fixing that.

IMU:  The Marble HD2 and Husky use the Lord Microstrain 3DM-GX5-25 IMU/AHRS and we derived the IMU noise model using the datasheet of this IMU so our solutions in simulation (for SLAM and so forth) would match the ones on our hardware vehicle. We certainly can change the model to be the same as the X1-X4, it is just convenient to have our simulation vehicle specs match the one in lab.

GIMBAL: The gimbal works only if you build a custom version of ros_ign bridge as there wasn't support for conversion from float64 to doubles. Are there no topics named "/X1/pan_tilt/pan_rate_cmd_double" or "/X1/pan_tilt/tilt_rate_cmd_double" then? I ran our vehicle models and  was unable to reproduce this. The topics listed above were there each time. Perhaps our local branches have changes that have since fixed this issue. `test_gimbal.sh` was able to work locally as a result.

We have some extenuating circumstances with access to our lab due to covid-19 so the 1m/s and 60 minute endurance are acceptable drop ins for now. In some previous email back and forths with osrf help they said the values wouldn't be needed because of this so I'm guessing the 1m/s and 60 minute endurance was the compromise you guys came up with. Thanks again for the flexibility during this time. 

I have a meeting later today with the team so we'll discuss these issues and submit a pull request to update our submission by tomorrow hopefully. Thanks!

**Comment by acschang** [comment]
Thank you for making the requested changes to the IMU characteristics and vehicle mass profile.

The RPLIDAR is mounted in such a way that the most rear-facing returns are blocked.  This is acceptable given your clarification on the mounting.

The gimbal test script is functional.  As noted in the `specifications.md`, the simulator is not adhering to the tilt limits in the `model.sdf`.

File: submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+      args="$(arg sensor_prefix)/camera_down/points@sensor_msgs/PointCloud2[ignition.msgs.PointCloudPacked">
+      <remap from="$(arg sensor_prefix)/camera_down/points" to="front_down_rgbd_camera/depth/points"/>
+    </node>
+    <node
+      pkg="ros_ign_image"
+      type="image_bridge"
+      name="ros_ign_image_depth_down"
+      args="$(arg sensor_prefix)/camera_down/depth_image">
+      <remap from="$(arg sensor_prefix)/camera_down/depth_image" to="front_down/depth"/>
+    </node>
```
Comment (iche033) [inline submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch:137]: here are the optical frame publisher nodes for `front_down` rgbd camera:

```
    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher_image_front_down">
      <remap from="input/image" to="front_down/image_raw" />
      <remap from="output/image" to="front_down/optical/image_raw" />
      <remap from="input/camera_info" to="front_down/camera_info" />
      <remap from="output/camera_info" to="front_down/optical/camera_info" />
    </node>
    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher_depth_down"
      args="depth">
      <remap from="input/image" to="front_down/depth" />
      <remap from="output/image" to="front_down/optical/depth" />
    </node>
```

File: submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      args="$(arg gimbal_sensor_prefix)/camera_pan_tilt/points@sensor_msgs/PointCloud2[ignition.msgs.PointCloudPacked">
+      <remap from="$(arg gimbal_sensor_prefix)/camera_pan_tilt/points" to="pan_tilt_rgbd_camera/depth/points"/>
+    </node>
+    <node
+      pkg="ros_ign_image"
+      type="image_bridge"
+      name="ros_ign_image_depth_pan_tilt"
+      args="$(arg gimbal_sensor_prefix)/camera_pan_tilt/depth_image">
+      <remap from="$(arg gimbal_sensor_prefix)/camera_pan_tilt/depth_image" to="pan_tilt/depth"/>
+    </node>
```
Comment (iche033) [inline submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch:167]: here are the optical frame publisher nodes for `pan_tilt` rgbd camera:

```
    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher_image_pan_tilt">
      <remap from="input/image" to="pan_tilt/image_raw" />
      <remap from="output/image" to="pan_tilt/optical/image_raw" />
      <remap from="input/camera_info" to="pan_tilt/camera_info" />
      <remap from="output/camera_info" to="pan_tilt/optical/camera_info" />
    </node>
    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher_depth_pan_tilt"
      args="depth">
      <remap from="input/image" to="pan_tilt/depth" />
      <remap from="output/image" to="pan_tilt/optical/depth" />
    </node>
```

File: subt_ros/launch/competition_init.launch [hunk 3]
```diff
@@ -18,10 +18,10 @@
     type="subt_ros_relay"
     name="subt_ros_relay"/>
 
-  <node
+  <!--<node
     pkg="subt_ros"
     type="bridge_logger"
-    name="bridge_logger"/>
+    name="bridge_logger"/>-->
```
Comment (iche033) [inline subt_ros/launch/competition_init.launch:24]: can you re-enable the `bridge_logger`?

File: submitted_models/marble_hd2_sensor_config_1/launch/spawner.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        <left_joint>front_middle_left_wheel_joint</left_joint>
+        <left_joint>rear_middle_left_wheel_joint</left_joint>
+        <left_joint>rear_left_wheel_joint</left_joint>
+        <right_joint>front_right_wheel_joint</right_joint>
+        <right_joint>front_middle_right_wheel_joint</right_joint>
+        <right_joint>rear_middle_right_wheel_joint</right_joint>
+        <right_joint>rear_right_wheel_joint</right_joint>
+        <wheel_separation>#{0.525}</wheel_separation>
+        <wheel_radius>0.129</wheel_radius>
+        <topic>/model/#{_name}/cmd_vel_relay</topic>
```
Comment (iche033) [inline submitted_models/marble_hd2_sensor_config_1/launch/spawner.rb:26]: I think velocity and acceleration limits may now be required. For example X1 uses these params

```
      <min_velocity>-1</min_velocity>
      <max_velocity>1</max_velocity>
      <min_acceleration>-3</min_acceleration>
      <max_acceleration>3</max_acceleration>
```
 maybe @acschang  can provide feedback on appropriate values for this vehicle.

**Comment by iche033** [comment]
I also noticed that this model does not support dropping breadcrumbs. If the functionality is allowed on this vehicle then the necessary topics and plugins need to be added.

File: submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch (hunk 2 above)
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch:167]: Should I remove the following?
```
    <node
      pkg="ros_ign_image"
      type="image_bridge"
      name="ros_ign_image_depth_pan_tilt"
      args="$(arg gimbal_sensor_prefix)/camera_pan_tilt/depth_image">
      <remap from="$(arg gimbal_sensor_prefix)/camera_pan_tilt/depth_image" to="pan_tilt/depth"/>
    </node>
```

File: submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch (hunk 1 above)
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch:137]: Should I remove the following image bridge then that was previously used for depth?
```

    <node
      pkg="ros_ign_image"
      type="image_bridge"
      name="ros_ign_image_depth_down"
      args="$(arg sensor_prefix)/camera_down/depth_image">
      <remap from="$(arg sensor_prefix)/camera_down/depth_image" to="front_down/depth"/>
    </node>
```

File: submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      args="$(arg sensor_prefix)/camera_front/points@sensor_msgs/PointCloud2[ignition.msgs.PointCloudPacked">
+      <remap from="$(arg sensor_prefix)/camera_front/points" to="front_facing_rgbd_camera/depth/points"/>
+    </node>
+    <node
+      pkg="ros_ign_image"
+      type="image_bridge"
+      name="ros_ign_image_depth_front"
+      args="$(arg sensor_prefix)/camera_front/depth_image">
+      <remap from="$(arg sensor_prefix)/camera_front/depth_image" to="front_facing/depth"/>
+    </node>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_1/launch/vehicle_topics.launch:107]: Should I remove the following image bridge then that was previously used for depth?
```
    <node
      pkg="ros_ign_image"
      type="image_bridge"
      name="ros_ign_image_depth_front"
      args="$(arg sensor_prefix)/camera_front/depth_image">
      <remap from="$(arg sensor_prefix)/camera_front/depth_image" to="front_facing/depth"/>
    </node>
```

File: subt_ros/launch/competition_init.launch (hunk 3 above)
Comment (bfotheri) [reply subt_ros/launch/competition_init.launch:24]: Done.

**Comment by bfotheri** [comment]
A pull request from the realdealneil fork of subt to branch model_pr_427 has been made with the appropriate updates based on the feedback here. Once that pull request is accepted this pull request the changes requested will be shown.

**Comment by iche033** [comment]
updated gimbal control and limits to be consistent with the configuration in pull request #384. 7326422

**Comment by nkoenig** [comment]
I'm merging this PR into the cave_feature_release3 branch. This robot will be available when the feature release branch is merged into master.

**Pull request merged**